### PR TITLE
Add inject_hyperparams wrapper

### LINKED ISSUE #20: Schedules for Non-Learning Rate Hyper-parameters
Is it possible to have schedules for hyper-parameters such as momentum? 

For example, onecycle momentum from this paper: https://arxiv.org/abs/1708.07120. There are also benefits to using schedules for parameters like epsilon (as a trust-region / damping parameter), highlighted by this blog post: http://zna.do/epsilon.

### PR DESCRIPTION
**Pull request by n2cholas**
Closes #20.

Adds schedule_hyperparams wrapper to schedule non-learning rate hyperparameters.

### CODE REVIEW

**Comment by mtthss** [comment]
I think this should live in `schedules.py` rather than in `wrappers.py`

**Comment by n2cholas** [comment]
> I think this should live in `schedules.py` rather than in `wrappers.py`

Makes sense, done! I needed to change `scale_by_schedule(step_size_fn: schedule.Schedule)` to `scale_by_schedule(step_size_fn: Callable[[chex.Numeric], chex.Numeric])` to break a circular import between `schedule.py` and `transform.py`. Let me know if you'd prefer I instead put the `Schedule` typedef somewhere common (perhaps `utils.py`?) and import it.

**Comment by 8bitmp3** [comment]
@n2cholas thanks for https://github.com/deepmind/optax/issues/20 I've only just seen it, this is awesome

cc @cgarciae @rwightman @RobertTLange @fehiepsi @KristianHolsheimer @rlouf

**Comment by mtthss** [comment]
> I needed to change scale_by_schedule(step_size_fn: schedule.Schedule)

That's fine since the functionality of `scale_by_schedule` is subsumed by your new component,
so we will slowly deprecate it anyway

File: optax/_src/schedule.py [hunk 1]
```diff
@@ -345,3 +347,64 @@ def cosine_onecycle_schedule(
       peak_value / div_factor,
       {int(pct_start * transition_steps): div_factor,
        int(transition_steps): 1. / (div_factor * final_div_factor)})
+
+
+class ScheduleHyperParamsState(transform.OptState):
```
[the remark this answers is not in the document]
Comment (mtthss) [reply optax/_src/schedule.py:352]: +1 to `ScheduleHyperparamsState`

[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/schedule.py]: Good catch, done!

[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/schedule.py]: Good point, thanks for catching this!

**Comment by rwightman** [comment]
I really like the idea.

Since naming things is always so much fun, it's either 'const' or 'scheduled' so not 100% on schedule_hyperparams as the wrapper name. What about `bind_hyperparams`?

Also, once this is in there, for my use I can't see myself ever not wanting this enabled, It'd probably either apply to every opt I use in a factory fn or write my own alias' that included the wrapper. Is the plan to have this fn in all current opt alias fns or to always have the user do the wrapping?

**Comment by n2cholas** [comment]
> Since naming things is always so much fun, it's either 'const' or 'scheduled' so not 100% on schedule_hyperparams as the wrapper name. What about `bind_hyperparams`?

I like `bind_hyperparams`!

> Also, once this is in there, for my use I can't see myself ever not wanting this enabled, It'd probably either apply to every opt I use in a factory fn or write my own alias' that included the wrapper. Is the plan to have this fn in all current opt alias fns or to always have the user do the wrapping?
> 
> EDIT Once I start using this, even if I had cases where everything was constant, I'd still use the wrapper to keep the interfaces consistent (ie accessing the hparams for logging, etc).

We initially planned to have the user do the wrapping, but you raise a good point. To add on to what you said, the current aliases support `ScalarOrSchedule` for the `learning_rate`, so it would also be nice for them to support `ScalarOrSchedule` for the rest of the hparams too.

@mtthss any thoughts?

**Comment by avital** [comment]
Apologies if this question is out of scope, or if I'm missing something (I probably am!) but does this approach only let you define schedules that are solely a function of the optimization step number? What if you want to dynamically choose, e.g. a different learning rate (or other hyperparameter) for each step of optimization based on some other factor (say the scale of the gradients of the previous optimization step or something).

This is a bit of a hypothetical question perhaps, as I don't have a concrete motivating factor, but in flax.optim we simply let the user specify values for any optimizer hyperparameter for each step of training explicitly.

/cc @jheek

**Comment by mtthss** [comment]
> To add on to what you said, the current aliases support ScalarOrSchedule for the learning_rate, so it would also be nice for them to support ScalarOrSchedule for the rest of the hparams too.

I think this CL actually subsumes the need to support `ScalarOrSchedule` in the individual transforms and aliases
I would make it *the* way of scheduling hyper-parameters
and remove the `ScalarOrSchedule` option all together

so that we don't need to worry at all about that in individual transforms and optimisers

**Comment by mtthss** [comment]
> I like `bind_hyperparams`

`bind` kind of makes it feel like they are fixed,
while this actually makes them dynamic

how about `dynamic_hyperparams`?

**Comment by mtthss** [comment]
> does this approach only let you define schedules that are solely a function of the optimization step number? 

Excellent question @avital! Worth discussing!

My immediate suggestion would be to keep this wrapper specific to schedules that are a function of the number of updates.

The reason is that, because of the composable nature of optax's GradientTransformations,
doing more arbitrary manual learning_rate rescalings is always possible by just
defining a custom transform that scales according to whatever you want (e.g. some stats of the previous updates)
and chaining it to the rest of the optimiser

Suppose I want to scale by the number of non zero gradients in previous steps
(this is not meant to be a smart thing to do, its just a random example)

```
def my_custom_scaler():

  def init_fn():
    return jnp.zeros((), dtype=jnp.float32)
  
  def update_fn(updates, state):
    scaled_updates = jax.tree_map(lambda t: state.num_zeros * t, updates)
    new_state  = sum(jax.tree_map(jnp.count_nonzero, jax.tree_leaves(updates)) 
    return scaled_updates, new_state

return GradientTransformation(init=init_fn, update=update_fn)
```
```

**Comment by mtthss** [comment]
@avital, just to clarify this is not to say that the one you suggest isn't an interesting use case,
but my gut feeling is that for that we might want to think a bit longer about the best approach
while this `wrapper` seems already very useful in its own right

**Comment by rwightman** [comment]
@mtthss `inject` works, less enthused about `dynamic` for same reasons that `schedule` isn't ideal in my mind. For me, `bind` doesn't imply the objects/values being bound are static, just that the grouping, the set of objects/values being bound are. If some of the bound values are callables, to me it strongly suggests that values returned could be 'dynamic'. The challenge with naming, everyone's semantics are slightly different.

Re @avital comment, definitely a valid use case, but once a param is determined by the optimization process is it a hyperparameter anymore? It seems the schedule/initial value as hparam would still be used to modulate or control the injection but the value itself is now part of the optimizer state...

**Comment by mtthss** [comment]
Let's go with `inject_hyperparams` then :)

**Comment by n2cholas** [comment]
@mtthss @rwightman Both `inject` and `bind` sound good to me, though I slightly prefer `bind`. 


> This is a bit of a hypothetical question perhaps, as I don't have a concrete motivating factor, but in flax.optim we simply let the user specify values for any optimizer hyperparameter for each step of training explicitly.

@avital I think your concern is valid and not hypothetical at all, for example, this is precisely the functionality you'd need to implement reducing the learning rate upon your loss/accuracy plateauing (which is pretty common). `flax.optim`'s approach wouldn't be directly applicable to Optax due to the way chaining transforms work (it's not clear to me if there's a clean way to pass the right hyper parameters to the right transforms)

@mtthss pointed out one possible approach, but below is approximately what I would do to dynamically change arbitrary hyperparameters (e.g. beta_1 for adam) during training:

```python
optim_factory = functools.partial(optax.adam, learning_rate=0.1, ...)

def train_step(batch, state):
    ...
    b1 = get_custom_b1(...)
    updates = optim_factory(b1=b1).update(grads, state.opt_state, params)
    ...
```

This pattern also works fine with the wrapper proposed in this PR:

```python
optim_factory = functools.partial(optax.schedule_hyperparams,  # name will be different
                                  inner_factory=optax.adam, 
                                  learning_rate=optax.exponential_decay(...), 
                                  ...)  
```
The use of `functools.partial` may be a bit complicated for new users, so there is definitely room for improvement. But perhaps an example with this pattern in the repo (e.g. for reducing the learning rate upon plateau) would be enough to familiarise people with it. Any thoughts on this?

**Comment by n2cholas** [comment]
Here's a potential alternative API (similar to the `optax.partial` @mtthss suggested). Currently the API is:

```python
custom_sgd = optax.inject_hyperparams(
      optax.sgd,
      learning_rate=optax.exponential_decay(...),
      momentum=optax.piecewise_constant_schedule(...),
      nesterov=True)
```

I was thinking this could be changed to:

```python
custom_sgd = optax.inject_hyperparams(optax.sgd)(
      learning_rate=optax.exponential_decay(...),
      momentum=optax.piecewise_constant_schedule(...),
      nesterov=True)
```

This way, `inject_hyperparams` can also function as a decorator, allowing us and users to easily wrap custom aliases (like @rwightman suggested):

```python
@optax.inject_hyperparams
def sgd(learning_rate: ScalarOrSchedule,
        momentum: ScalarOrSchedule = 0.,
        nesterov: bool = False) -> GradientTransformation:
  return optax.chain(
      optax.trace(decay=momentum, nesterov=nesterov),
      optax.scale(learning_rate),
  )
```

The implementation would look something like:

```python
def inject_hyperparams(
    inner_factory: Callable[..., GradientTransformation],
) -> Callable[..., GradientTransformation]:

  def transform_wrapper(**hyperparams: Union[Schedule, Any]) -> GradientTransformation:

    def schedule_fn(count):
        return {k: (f(count) if callable(f) else f) for k, f in hyperparams.items()}
    def init_fn(params): ...
    def update_fn(updates, state, params=None): ...

    return GradientTransformation(init_fn, update_fn)

  return transform_wrapper
```

Any thoughts on this?

**Comment by mtthss** [comment]
I think this is great!
Very clean and I love that it doubles as decorator

**Comment by rwightman** [comment]
@n2cholas I like it!

**Comment by n2cholas** [comment]
> What if you want to dynamically choose, e.g. a different learning rate (or other hyperparameter) for each step of optimization based on some other factor (say the scale of the gradients of the previous optimization step or something).

A potential idea to address this use case: what if we allowed users to change the OptState `hyperparams` dict themselves for non-scheduled values? e.g.

```python
optim = inject_hyperparams(optax.sgd)(
    learning_rate=0.1, 
    momentum=optax.piecewise_constant_schedule(...))
opt_state = optim.init(params)

def train_step(batch, opt_state):
  ...
  opt_state.hyperparams['learning_rate'] = get_custom_lr(...)
  updates, state = optim.update(grads, opt_state)
  ...
```

This would change the implementation so only the scheduled hparams are updated:

```python
def schedule_fn(count):  # now only returns scheduled hparams
  return {k: f(count) for k, f in hyperparams.items() if callable(f)}

def init_fn(params):
  count = jnp.zeros([], jnp.int32)
  hparams = {**hyperparams, **schedule_fn(count)}  # merge constants and scheduled
  return InjectHyperparamsState(
      count=count,
      hyperparams=hparams,
      inner_state=inner_factory(**hparams).init(params))

def update_fn(updates, state, params=None):
  count_inc = utils.safe_int32_increment(state.count)
  hparams = {**state.hyperparams, schedule_fn(count_inc)}  # only override scheduled hparams
  updates, inner_state = inner_factory(**hparams).update(
      updates, state.inner_state, params)
  return updates, InjectHyperparamsState(
      count=count_inc,
      hyperparams=hparams,
      inner_state=inner_state)
```

It adds a _bit_ more complexity, but seems worth it for the additional functionality in my view. Any thoughts on this @mtthss @avital?

**Comment by mtthss** [comment]
Nice! I think we are really getting somewhere!
Thanks for the many iterations on the design.

**Comment by avital** [comment]
@n2cholas first of of all, thank you for being here as the expert on both flax.optim and Optax :)

The pattern you proposed, i.e.

```py
def train_step(batch, opt_state):
  ...
  opt_state.hyperparams['learning_rate'] = get_custom_lr(...)
  updates, state = optim.update(grads, opt_state)
  ...
```

does indeed seem like it replicates the behavior of the `hyper_param_overrides` argument to [`flax.optim.Optimizer.apply_gradient`](https://flax.readthedocs.io/en/latest/flax.optim.html#flax.optim.Optimizer.apply_gradient)

/cc @andsteing @jheek

So I think that resolves my concern.

**Comment by mtthss** [comment]
@andsteing @jheek do you have some final comments?
If not, I think we can proceed with the final proposal by @n2cholas

**Comment by andsteing** [comment]
This is neat, I like it!

The common case with the schedules is covered elegantly, and while Avital's `optimization based on some other factor` case is a bit more delicate (e.g. no error if misspelled or overwritten by a schedule) I think that's fine since it seems quite an edge/expert usecase to me.

**Comment by n2cholas** [comment]
Thanks everyone!  I've implemented these changes and updated `aliases.py` to use this decorator for all the aliases. The implementation was slightly more nuanced than proposed: we cannot put boolean flags in the `InjectHyperparamsState` otherwise it won't work with JIT (so only numeric hyper parameters are stored in the `hyperparams` dict).

**Comment by n2cholas** [comment]
One drawback of the current design is there is no way of telling schedules apart from other callable arguments (e.g. the new callable mask allowed for `optax.masked` (#60)). 

One possible solution is to add an `ignore` argument to `inject_hyperparams`, which accepts a list of hyperparameter names to ignore for scheduling. Usage example:

```python
optim = inject_hyperparams(custom_optimizer, ignore=['mask'])(
    learning_rate=schedule(...), mask=mask_fn)
```
Decorator example:

```python
@functools.partial(inject_hyperparams, ignore=['mask'])
def my_optimizer(learning_rate: float, mask: Callable):
  ...
```

Any feedback or suggestions on this?

**Comment by andsteing** [comment]
Yeah the assumption `is callable -> must be schedule` seems brittle, especially with a framework like Optax where it feels natural to use higher order functions for all kind of purposes.

What about making schedule creating functions return some typed function?

This could be a `NamedTuple` with a single field that is the scheduling function

```python
class Schedule(NamedTuple):
  schedule: Callable[[int], float]
```

Or a wrapper class 

```python
class Schedule:
  def __init__(self, fun):
    self.fun = fun
  def __call__(self, step):
    return self.fun(step)
```

**Comment by mtthss** [comment]
I am a bit reluctant to require people to put schedules in a NamedTuple with a field schedule because several people don't use only the canonical schedules in optax, and they also define their own custom schedules, and its nice that they can do so inline in a very lightweight manner.

I don't have yet though a concrete alternative to suggest for how to solve the issue of the wrapper avoiding considering all callable as schedules...

**Comment by mtthss** [comment]
Overall I think it's reasonable that by default `inject_hyperparams` assumes that callables are schedules. You just need to provide a simple override mechanism.

You already proposed one, but I would rename `ignore` to `static_args` (similarly to `static_argnums` in `jit`)

**Comment by n2cholas** [comment]
> Overall I think it's reasonable that by default `inject_hyperparams` assumes that callables are schedules. You just need to provide a simple override mechanism.
> 
> You already proposed one, but I would rename `ignore` to `static_args` (similarly to `static_argnums` in `jit`)

Makes sense, I agree. JAX users should be familiar with this pattern given it matches JIT, VMAP, etc. I like the name `static_args`!

**Review by mtthss — APPROVED** [review]

**Comment by mtthss** [comment]
@n2cholas  why is the `asarray` call needed in:
```
  elif isinstance(value, (int, float, chex.Array)):
    numeric_hps[name] = jnp.asarray(value)
```
within the `inject` wrapper?

**Comment by mtthss** [comment]
I changed to code to avoid the `jnp.asarray` in that line
and instead do

```
hparams = {
    k: jnp.asarray(_convert_floats(v, dtype))
    for k, v in numeric_hps.items()}
```

In the `init_fn`

**Comment by n2cholas** [comment]
> @n2cholas why is the `asarray` call needed in:
> can I just get rid of it since the content of `numeric_hps` already gets passed to `_convert_floats` later on?

Integer hyperparameters also need to be converted to arrays.

> I changed to code to avoid the `jnp.asarray` in that line
> and instead do
> 
> ```
> hparams = {
>     k: jnp.asarray(_convert_floats(v, dtype))
>     for k, v in numeric_hps.items()}
> ```
> 
> In the `init_fn`

This works, thanks for making the change!

**Comment by mtthss** [comment]
I started merging the addition of the `inject_hyperparams` wrapper,
i will follow up soon with the second part (enabling it by default in all optimisers in `alias.py`)

**Comment by n2cholas** [comment]
That's great, thank you!
